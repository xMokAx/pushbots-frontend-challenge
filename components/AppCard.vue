<template>
  <v-card class="pa-2 pa-sm-4">
    <div class="d-flex justify-space-between">
      <div class="d-flex">
        <v-avatar tile size="64" class="d-none d-sm-block">
          <img :src="app.icon" :alt="app.title" />
        </v-avatar>
        <div>
          <v-card-title class="pt-0 pl-2">{{ app.title }}</v-card-title>
          <v-card-subtitle class="pl-2">
            <v-icon small>mdi-cellphone-iphone</v-icon>
            <span class="mr-2">{{ app.totalUsers }} users</span>
            <v-icon v-for="platform in platforms" :key="platform" small>
              {{ platform }}
            </v-icon>
          </v-card-subtitle>
        </div>
      </div>
      <div class="d-flex pt-2">
        <v-btn
          v-for="{ icon, label } in topButtons"
          :key="icon"
          :aria-label="label"
          text
          icon
        >
          <v-icon>{{ icon }}</v-icon>
        </v-btn>
      </div>
    </div>
    <Chart :labels="chartLabels" :data="chartValues" class="chart" />

    <v-btn
      v-for="{ icon, label } in bottomButtons"
      :key="icon"
      :aria-label="label"
      text
      icon
    >
      <v-icon>{{ icon }}</v-icon>
    </v-btn>
  </v-card>
</template>

<script>
import Chart from './Chart'
export default {
  components: {
    Chart
  },
  props: {
    app: {
      type: Object,
      required: true
    }
  },
  data: () => ({
    topButtons: [
      {
        icon: 'mdi-decagram',
        label: 'star app'
      },
      {
        icon: 'mdi-share',
        label: 'share app'
      },
      {
        icon: 'mdi-dots-vertical',
        label: 'more options'
      }
    ],
    bottomButtons: [
      {
        icon: 'mdi-send',
        label: 'send notification'
      },
      {
        icon: 'mdi-update',
        label: 'update app'
      },
      {
        icon: 'mdi-cellphone-iphone',
        label: 'mobile devices'
      },
      {
        icon: 'mdi-account-group',
        label: 'all devices'
      },
      {
        icon: 'mdi-settings',
        label: 'app settings'
      },
      {
        icon: 'mdi-chart-areaspline-variant',
        label: 'app statistics'
      },
      {
        icon: 'mdi-twitter',
        label: 'tweet'
      }
    ]
  }),
  computed: {
    chartLabels() {
      return this.app.chartData.map((d) => d.label)
    },
    chartValues() {
      return this.app.chartData.map((d) => d.value)
    },
    platforms() {
      return Object.entries(this.app.platforms)
        .map(([p, c]) => {
          if (c) {
            let icon
            switch (p) {
              case 'ios':
                icon = 'apple'
                break
              case 'chrome':
                icon = 'google-chrome'
                break
              case 'safari':
                icon = 'apple-safari'
                break
              default:
                icon = p
            }
            return `mdi-${icon}`
          } else {
            return false
          }
        })
        .filter((icon) => !!icon)
    }
  }
}
</script>

<style scoped>
.legend {
  border: 1px solid white;
}
.chart {
  max-height: 120px;
}
</style>
