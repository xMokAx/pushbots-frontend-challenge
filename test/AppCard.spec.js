import Vue from 'vue'
import Vuetify from 'vuetify'
import { mount } from '@vue/test-utils'
import AppCard from '../components/AppCard'

Vue.use(Vuetify)

const propsData = {
  app: {
    id: '28d1e966-7508-4ee9-a594-74eb146b7671',
    title: 'fumbling needy knife',
    icon: 'https://picsum.photos/50/50',
    totalUsers: 62336,
    platforms: {
      ios: true,
      android: true,
      chrome: true,
      firefox: true,
      opera: true,
      safari: true
    },
    chartData: [
      {
        label: '2019-11-13',
        value: 27
      },
      {
        label: '2019-11-14',
        value: 121
      },
      {
        label: '2019-11-15',
        value: 938
      },
      {
        label: '2019-11-16',
        value: 517
      },
      {
        label: '2019-11-17',
        value: 748
      },
      {
        label: '2019-11-18',
        value: 845
      },
      {
        label: '2019-11-19',
        value: 693
      },
      {
        label: '2019-11-20',
        value: 324
      }
    ]
  }
}

describe('AppCard', () => {
  it('Should match snapshot', () => {
    const wrapper = mount(AppCard, {
      propsData
    })
    expect(wrapper.html()).toMatchSnapshot()
  })
})
