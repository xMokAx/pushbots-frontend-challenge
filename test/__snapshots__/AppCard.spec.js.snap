// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppCard Should match snapshot 1`] = `"<div class=\\"pa-2 pa-sm-4 v-card v-sheet theme--light\\"><div class=\\"d-flex justify-space-between\\"><div class=\\"d-flex\\"><div class=\\"v-avatar d-none d-sm-block v-avatar--tile\\" style=\\"height: 64px; min-width: 64px; width: 64px;\\"><img src=\\"https://picsum.photos/50/50\\" alt=\\"fumbling needy knife\\"></div> <div><div class=\\"v-card__title pt-0 pl-2\\">fumbling needy knife</div> <div class=\\"v-card__subtitle pl-2\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-cellphone-iphone theme--light\\" style=\\"font-size: 16px;\\"></i> <span class=\\"mr-2\\">62336 users</span> <i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-apple theme--light\\" style=\\"font-size: 16px;\\"></i><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-android theme--light\\" style=\\"font-size: 16px;\\"></i><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-google-chrome theme--light\\" style=\\"font-size: 16px;\\"></i><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-firefox theme--light\\" style=\\"font-size: 16px;\\"></i><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-opera theme--light\\" style=\\"font-size: 16px;\\"></i><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-apple-safari theme--light\\" style=\\"font-size: 16px;\\"></i></div></div></div> <div class=\\"d-flex pt-2\\"><button type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default\\" aria-label=\\"star app\\"><span class=\\"v-btn__content\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-decagram theme--light\\"></i></span></button><button type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default\\" aria-label=\\"share app\\"><span class=\\"v-btn__content\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-share theme--light\\"></i></span></button><button type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default\\" aria-label=\\"more options\\"><span class=\\"v-btn__content\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-dots-vertical theme--light\\"></i></span></button></div></div> <div class=\\"chart\\"><canvas id=\\"line-chart\\" width=\\"0\\" height=\\"0\\" style=\\"display: block;\\" class=\\"chartjs-render-monitor\\"></canvas></div> <button type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default\\" aria-label=\\"send notification\\"><span class=\\"v-btn__content\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-send theme--light\\"></i></span></button><button type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default\\" aria-label=\\"update app\\"><span class=\\"v-btn__content\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-update theme--light\\"></i></span></button><button type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default\\" aria-label=\\"mobile devices\\"><span class=\\"v-btn__content\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-cellphone-iphone theme--light\\"></i></span></button><button type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default\\" aria-label=\\"all devices\\"><span class=\\"v-btn__content\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-account-group theme--light\\"></i></span></button><button type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default\\" aria-label=\\"app settings\\"><span class=\\"v-btn__content\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-settings theme--light\\"></i></span></button><button type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default\\" aria-label=\\"app statistics\\"><span class=\\"v-btn__content\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-chart-areaspline-variant theme--light\\"></i></span></button><button type=\\"button\\" class=\\"v-btn v-btn--flat v-btn--icon v-btn--round v-btn--text theme--light v-size--default\\" aria-label=\\"tweet\\"><span class=\\"v-btn__content\\"><i aria-hidden=\\"true\\" class=\\"v-icon notranslate mdi mdi-twitter theme--light\\"></i></span></button></div>"`;
